packages/gatsby/src/schema/infer-graphql-type.js

```diff
@@ -256,21 +257,23 @@ function inferFromFieldName(value, selector, types): GraphQLFieldConfig<*, *> {
     let type
     // If there's more than one type, we'll create a union type.
     if (fields.length > 1) {
-      type = new GraphQLUnionType({
-        name: createTypeName(
-          `Union_${key}_${fields
+      type = unionTypes.find(type =>
```

Comment:
let's make `unionTypes` an object/map so we don't need to find type in array.

map keys could be name of type, so first thing here is to generate typeName:
```
const typeName = `Union_${key}_${fields.map(f => f.name).sort().join(`__`)
```
check if that typeName is already in our map and return it of it is

if it's not create type and set our new type in our map

Reply:
I've made `unionTypes` an object, so union type instances can be shared on a per-key basis. The value still needs to be an array though (and we still need to use `.find` on it), because of this use-case:

- A node type contains a union type on the field with the key `content`;
- Another node type also contains a union type on the `content` field, but it contains different node types for the children.

In this case, we do _not_ want to share the union type. We want to verify that it actually contains the same node types.

In case you hadn't already noticed, I've changed the name of the union type to exclude all fields names. This way the union type name doesn't change when you the data structure for the children change (in case of WordPress: adding or removing an ACF layout).

Reply:
Right, that's why object/map key would need to include field type names and not just field `key`.

Reply:
I've been thinking about this, but I think there's an issue with using the field type names as a key. The reason behind this, is that the [schema is updated](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/bootstrap/index.js#L377) when bootstrapping Gatsby. On bootstrap, the schema is re-created.

If we are using the field _names_ as a key, the _same_ union type instance that is created during initial schema creation, is also used when the schema is updated. The problem here is that the union type defines which child node types it has, which are in turn registered as fields during schema creation (in `GraphQLSchema`).

Since the 'old' union type (and its child node types) are used, the schema tries to create the child node type for both the old union type _and_ the new child node type during schema creation. Since GraphQL's JS implementation checks for type names being the exact same type (and compares the object itself instead of just the names), [this error](https://github.com/graphql/graphql-js/blob/master/src/type/schema.js#L278-L284) would be thrown.

This is why comparing the objects using `_.isEqual` _does_ work, since they are not equal and a new union type is actually created when updating the schema (which it should). I think we have two options:

1. Resetting the `unionTypes` map when starting to construct schema (as you proposed for `seenNames`)
2. Keep comparing types using `.find` and `_.isEqual`

Do you have any other ideas? Which option has your preference?

Reply:
We can clear that map when we start (re)building schema (in `schema/index.js`)

Reply:
Thanks for your reply. I've updated the PR and converted to a `Map` for ease of clearing the union types map.